x-pack/plugins/proxy/server/proxy.ts

```diff
@@ -0,0 +1,291 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { Agent as HTTPSAgent } from 'https';
+import { Agent as HTTPAgent } from 'http';
+import { URL } from 'url';
+
+import { Request, ResponseToolkit } from 'hapi';
+import { Observable, Subscription, from } from 'rxjs';
+import { first } from 'rxjs/operators';
+import { v4 } from 'uuid';
+import Wreck from 'wreck';
+
+import {
+  PluginInitializerContext,
+  ClusterClient,
+  Logger,
+  CoreStart,
+  CoreSetup,
+  Router,
+} from 'target/types/server';
+
+import { HttpServerSetup } from 'target/types/server/http/http_server';
+import { ProxyConfig } from './proxy_config';
+
+export enum RouteState {
+  Initializing,
+  Started,
+  Closed,
+}
+
+export interface RoutingNode {
+  type: string; // what are all the types this can be?
+  node: string;
+  state: RouteState;
+}
+
+interface LivenessNode {
+  lastUpdate: number;
+  lastChangedTime: number;
+}
+
+interface ClusterDoc {
+  nodes: {
```

Comment:
is not the same as 
```typescript
interface ClusterDoc {
  nodes: NodeList;
  routing_table: RoutingTable;
}
```
?

Reply:
`NodeList` and `RoutingTable` are both `Map<K,V>` -- the `ClusterDoc` is what elasticsearch returns, so it's a JSON structure and not a map.

I could remove the Maps and just make `NodeList` and `RoutingTable` into `interface { [key:string]: LivenessNode}`

Reply:
Up to you. It's optional. I'd prefer to have one common type here: Map, Record, POJO, whatever else.